kratos/containers/csr_matrix.h

```diff
@@ -111,21 +117,77 @@ class CsrMatrix
         });
     }
 
+    //move constructor
+    CsrMatrix(CsrMatrix<TDataType,TIndexType>&& rOtherMatrix)
+    {
+        rOtherMatrix.mIsOwnerOfData=false;
+        mIsOwnerOfData=true;
+
+        //swap the pointers to take owership of data
+        mpRowIndicesData = rOtherMatrix.mpRowIndicesData;
+        mpColIndicesData = rOtherMatrix.mpColIndicesData;
+        mpValuesVectorData = rOtherMatrix.mpValuesVectorData;
+
+        //here we assign the span
+        mRowIndices = rOtherMatrix.mRowIndices;
+        mColIndices = rOtherMatrix.mColIndices;
+        mValuesVector = rOtherMatrix.mValuesVector;
+        
+        mNrows = rOtherMatrix.mNrows;
+        mNcols = rOtherMatrix.mNcols;
+
+    }
 
     /// Destructor.
-    virtual ~CsrMatrix(){}
+    virtual ~CsrMatrix(){
+        // if(mpRowIndicesData != nullptr) 
+        //     delete [] mpRowIndicesData;
+        // if(mpColIndicesData != nullptr) 
+        //     delete [] mpColIndicesData;
+        // if(mpValuesVectorData != nullptr) 
+        //     delete [] mpValuesVectorData;
+
+        // mpRowIndicesData = nullptr;
+        // mpColIndicesData = nullptr;
+        // mpValuesVectorData = nullptr
+        AssignIndex1Data(nullptr,0);
+        AssignIndex2Data(nullptr,0);
+        AssignValueData(nullptr,0);        
```

Comment:
This is confusing, who deleted the data?

Reply:
the Assign functions replace the data and do the deletion internally. Actually this was one of the hard parts of the PR